src/main/jni/src/genomicsdb_GenomicsDBQuery.cc

```diff
@@ -0,0 +1,319 @@
+    
+    //org.genomicsdb.reader.GenomicsDBQuery$VariantCalls
+    INIT(java_VariantCalls_, static_cast<jclass>(env->NewGlobalRef(env->FindClass("org/genomicsdb/reader/GenomicsDBQuery$VariantCalls"))));
+    INIT(java_VariantCalls_init_default_,  env->GetMethodID(java_VariantCalls_, "<init>", "()V"));
+    INIT(java_VariantCalls_init_,  env->GetMethodID(java_VariantCalls_, "<init>", "(JJ)V"));
+
+    //org.genomicsdb.reader.GenomicDBQuery$VariantCall
+    INIT(java_VariantCall_, static_cast<jclass>(env->NewGlobalRef(env->FindClass("org/genomicsdb/reader/GenomicsDBQuery$VariantCall"))));
+    INIT(java_VariantCall_init_, env->GetMethodID(java_VariantCall_, "<init>", "(ILjava/lang/String;JJLjava/util/Map;)V"));
+
+    //org.genomicsdb.reader.GenomicsDBQuery$Pair<L, R>
+    INIT(java_Pair_, static_cast<jclass>(env->NewGlobalRef(env->FindClass("org/genomicsdb/reader/GenomicsDBQuery$Pair"))));
+    INIT(java_Pair_init_, env->GetMethodID(java_Pair_, "<init>", "(JJ)V"));
+    INIT(java_Pair_getStart_, env->GetMethodID(java_Pair_, "getStart", "()J"));
+    INIT(java_Pair_getEnd_, env->GetMethodID(java_Pair_, "getEnd", "()J"));
+    
+    is_initialized_ = true;
+  }
+}
+
+void JNI_OnUnload(JavaVM *vm, void *reserved) {
+  return;
+  std::lock_guard<std::mutex> guard(initializing_mutex_);
+  if (is_initialized_) {
+    JNIEnv *env;
+    if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_8) == JNI_OK) {
+      env->DeleteGlobalRef(java_Pair_);
+      env->DeleteGlobalRef(java_VariantCall_);
+      env->DeleteGlobalRef(java_VariantCalls_);
+      env->DeleteGlobalRef(java_HashMap_);
+      env->DeleteGlobalRef(java_ArrayList_);
+
+      is_initialized_ = false;
+    }
+  }
+}
+
+#if(0)
+# Useful for debugging JNI
+void get_class_name(JNIEnv *env, jobject obj) {
+  jclass cls = env->GetObjectClass(obj);
+  jmethodID mid = env->GetMethodID(cls, "getClass", "()Ljava/lang/Class;");
+  jobject clsObj = env->CallObjectMethod(obj, mid);
+  cls = env->GetObjectClass(clsObj);
+  mid = env->GetMethodID(cls, "getName", "()Ljava/lang/String;");
+  jstring strObj = (jstring)env->CallObjectMethod(clsObj, mid);
+  const char* str = env->GetStringUTFChars(strObj, NULL);
+  // Print the class name
+  printf("\nCalling class is: %s\n", str);
+  env->ReleaseStringUTFChars(strObj, str);
+}
+#endif
+
+std::vector<std::string> to_string_vector(JNIEnv *env, jobject arrayList) {
+  initialize(env);
+  jint size = env->CallIntMethod(arrayList, java_ArrayList_size_);
+  std::vector<std::string> result;
+  result.reserve(size);
+  for (jint i=0; i<size; i++) {
+    jstring element = static_cast<jstring>(env->CallObjectMethod(arrayList, java_ArrayList_get_, i));
+    auto element_cstr = env->GetStringUTFChars(element, nullptr);
+    result.emplace_back(element_cstr);
+    env->ReleaseStringUTFChars(element, element_cstr);
+    env->DeleteLocalRef(element);
+  }
+  return std::move(result);
```

Comment:
I don't think it helps here, does it? Generally, I have seen moves help in the caller function

Reply:
Yes, it is not necessary, return value optimization will do the right thing. Thanks.